### ISSUE TITLE
Pyramid 1.10.3 is much slower to initiate

### ISSUE THREAD

**Issue by Deimos** [Pylons/pyramid#3489]
Just updated to Pyramid 1.10.3 and it's now considerably slower to initiate. The main reason I notice this is because I'm using [WebTest](https://docs.pylonsproject.org/projects/webtest/en/latest/index.html) and my first test that uses it is now much slower than it used to be.

On Pyramid 1.10.2, the setup time for my first WebTest test averages about 0.7 seconds. After updating to Pyramid 1.10.3 (and changing nothing else), the setup time for the same test increases to about 3.8 seconds.

Similarly, starting up `pshell` is noticeably slower. I haven't tried to time it properly, but I can now easily count past 5 while waiting for it to start up, but on Pyramid 1.10.2 I can't even get to 3.

I can try to help investigate where the additional time is being spent if needed, but I figured you might already know which of the 1.10.3 changes could be the culprit.

**Comment by mmerickel** [comment]
You should check that other libraries are at the same versions as you test and then finally if that's all constant then check if the changes made to view_config to improve stackframe captures [1]. The diff in this release is ultra tiny, affecting only pserve (but you said you're using bootstrap, not pserve) and view_config.

https://github.com/Pylons/pyramid/compare/1.10.2...1.10.3

**Comment by Deimos** [comment]
Sorry, I shouldn't have used the word "bootstrap". I'll try to do some profiling and see if I can figure out where the extra time is being spent.

**Comment by mmerickel** [comment]
Gotcha, well I suspect it's because of the stackframe code in the diff if it is slower but it'd help if you could narrow down that it's at least not related to any third party dependency issues in your own project first.

**Comment by luhn** [comment]
Hmm, I figured there would be a performance regression with the new view_config, but nothing so dramatic.  venusian already inspects the frame, so I figured worst-case would be 2x slowdown.

Just spitballing, it could be that inspect.stack() calls getframeinfo for every frame in the stack, whereas venusian only calls getframeinfo for a single frame.

Chad, would you mind testing against https://github.com/luhn/pyramid/tree/performance-regression? <https://github.com/luhn/pyramid/tree/performance-regression?>  (`pip install --force https://github.com/luhn/pyramid/archive/performance-regression.zip` <https://github.com/luhn/pyramid/archive/performance-regression.zip%60>)

— Theron

**Comment by Deimos** [comment]
Alright, I did some quick profiling, with absolutely nothing changing except upgrading pyramid from 1.10.2 to 1.10.3. You're right, the difference does seem to be the stackframe code.

Here's the cProfile line for my app's init under 1.10.2:

```
   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    0.784    0.784 /opt/tildes/tildes/__init__.py:21(main)
```

Under 1.10.3, that turns into a much longer process, with most of the time taken up by frame-related calls:

```
   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    4.715    4.715 /opt/tildes/tildes/__init__.py:21(main)
     8257    0.034    0.000    4.282    0.001 /usr/lib/python3.7/inspect.py:1444(getframeinfo)
      111    0.000    0.000    4.145    0.037 /usr/lib/python3.7/inspect.py:1511(stack)
      111    0.012    0.000    4.144    0.037 /usr/lib/python3.7/inspect.py:1483(getouterframes)
      110    0.000    0.000    4.132    0.038 /opt/venvs/tildes/lib/python3.7/site-packages/pyramid/view.py:217(__init__)
      110    0.002    0.000    4.132    0.038 /opt/venvs/tildes/lib/python3.7/site-packages/pyramid/view.py:224(_get_info)
        4    0.000    0.000    3.901    0.975 /opt/venvs/tildes/lib/python3.7/site-packages/pyramid/config/__init__.py:811(scan)
        4    0.001    0.000    3.899    0.975 /opt/venvs/tildes/lib/python3.7/site-packages/venusian/__init__.py:20(scan)
     9772    1.408    0.000    3.343    0.000 /usr/lib/python3.7/inspect.py:714(getmodule)
     8257    0.044    0.000    2.894    0.000 /usr/lib/python3.7/inspect.py:760(findsource)
    17988    0.048    0.000    2.624    0.000 /usr/lib/python3.7/inspect.py:680(getsourcefile)
       61    0.000    0.000    2.106    0.035 /opt/tildes/tildes/views/decorators.py:13(ic_view_config)
```

1.10.3 spends 4.132 seconds in `/opt/venvs/tildes/lib/python3.7/site-packages/pyramid/view.py:217(__init__)`, but 1.10.2 doesn't even measure any time under that (it says 0.000).

GitHub isn't allowing me to attach the cProfile output files, but I could upload them or email them to you if you want to dig into it more with `pstats`. Let me know if there's anything else I can do to help.

**Comment by Deimos** [comment]
Sorry @luhn, I didn't see your comment until after I finished mine. I'll test that in a little while, thanks!

**Comment by bertjwregeer** [comment]
It looks like you are right @luhn: https://github.com/python/cpython/blob/74125a60b7a477451ff2b8385bfbce3fdaee8dbc/Lib/inspect.py#L1522

It calls getouterframes, on `sys._getframe(1)`. I'm not a huge fan of calling internal API's, but maybe in this case it is warranted :-(

**Comment by Deimos** [comment]
One more chunk of details since I was still poking around in `pstats`:

1.10.2 functions that spent at least 0.05s in tottime:

```
   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
    13376    0.519    0.000    0.519    0.000 {built-in method posix.stat}
     1278    0.135    0.000    0.135    0.000 {built-in method marshal.loads}
      161    0.130    0.001    0.130    0.001 {method 'execute' of 'psycopg2.extensions.cursor' objects}
    11512    0.101    0.000    0.101    0.000 {built-in method posix.lstat}
        1    0.100    0.100    0.100    0.100 {built-in method time.sleep}
      507    0.081    0.000    0.081    0.000 {built-in method io.open}
71194/70156    0.077    0.000    0.087    0.000 {built-in method builtins.hasattr}
        2    0.061    0.031    0.061    0.031 {built-in method _ffi.argon2_hash}
     1278    0.053    0.000    0.077    0.000 <frozen importlib._bootstrap_external>:914(get_data)
4057/3992    0.051    0.000    0.311    0.000 {built-in method builtins.__build_class__}
```

1.10.3:

```
   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
     9772    1.408    0.000    3.343    0.000 /usr/lib/python3.7/inspect.py:714(getmodule)
    37016    1.053    0.000    1.053    0.000 {built-in method posix.stat}
2698976/2697938    0.751    0.000    0.833    0.000 {built-in method builtins.hasattr}
  2687638    0.488    0.000    0.692    0.000 /usr/lib/python3.7/inspect.py:63(ismodule)
  3120305    0.252    0.000    0.265    0.000 {built-in method builtins.isinstance}
  2568866    0.251    0.000    0.253    0.000 {method 'get' of 'dict' objects}
        2    0.200    0.100    0.200    0.100 {built-in method time.sleep}
      161    0.155    0.001    0.156    0.001 {method 'execute' of 'psycopg2.extensions.cursor' objects}
     1278    0.135    0.000    0.135    0.000 {built-in method marshal.loads}
    11512    0.105    0.000    0.105    0.000 {built-in method posix.lstat}
      511    0.080    0.000    0.081    0.000 {built-in method io.open}
     1278    0.058    0.000    0.086    0.000 <frozen importlib._bootstrap_external>:914(get_data)
        2    0.058    0.029    0.058    0.029 {built-in method _ffi.argon2_hash}
4056/3991    0.052    0.000    0.468    0.000 {built-in method builtins.__build_class__}
```

2-3 million additional calls to `hasattr`, `ismodule`, `isinstance` and `dict.get` seem to causing a lot of the additional time.

**Comment by Deimos** [comment]
@luhn your branch seems to fix it. It adds an additional 110 calls to `_getframe` and `getframeinfo`, but only about 0.04s of additional time.

**Comment by bertjwregeer** [comment]
@Deimos thanks for verifying the fix works.

**Comment by Deimos** [comment]
No problem at all, thanks (everyone) for the quick responses and investigation. Do you think there might be anything from this that should be reported to the actual Python maintainers to improve the `inspect` module?

**Comment by mmerickel** [comment]
Thank you everyone for the quick turnaround. Great work! Fixed in #3491
